internal/auth/oidc/callback_url.go

```diff
@@ -0,0 +1,98 @@
+package oidc
+
+import (
+	"net/url"
+
+	"github.com/hashicorp/boundary/internal/auth/oidc/store"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/oplog"
+	"google.golang.org/protobuf/proto"
+)
+
+// DefaultCallbackUrlTableName defines the default table name for a CallbackUrl
+const DefaultCallbackUrlTableName = "auth_oidc_callback_url"
+
+// CallbackUrl defines an callback URL for an OIDC auth method. It is assigned
+// to an OIDC AuthMethod and updates/deletes to that AuthMethod are cascaded to
+// its CallbackUrls.
+//
+// see redirect_uri in the oidc spec:
+// https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest
+type CallbackUrl struct {
+	*store.CallbackUrl
+	tableName string
+}
+
+// NewCallbackUrl creates a new in memory callback assigned to an OIDC
+// AuthMethod.  It supports no options.
+//
+// For more info on oidc callbacks, see redirect_uri in the oidc spec:
+// https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest
+func NewCallbackUrl(authMethodId string, callback *url.URL) (*CallbackUrl, error) {
```

Comment:
What is the point of having a callback url?  I had expected this to be hard coded in the API and not something stored in the database.  For example, if I have auth method `oidcam_123` then the callback url would always be `<whatever the domain is>/v1/auth-methods/oidc_123:callback`.

Reply:
After an earlier discussion with @mgaffney, I believe we needed to support setting the domain portion for the callback and that the pathing would be statically set by Boundary.

Reply:
How would the API look? I think the answer to that determines what this method looks like.

Would it allow the full url to be provided and just error if the suffix was wrong?  Would it accept just the part before `/v1/auth-methods/amoidc_123` but append that before storage?  Would it accept/store/return only the first part and maybe have a different readonly field which appends the `/v1/auth-methods/amoidc_123` portion to what is entered?

Reply:
I sorted of "punted" on what portion would be stored in the db.  I know for certain that it must be a valid URL, so that's what I've implemented so far.  I'm not sure the repo should enforce what portion is stored vs what is "read-only" and appended for display.  The service layer would seem to own that responsibility in our implementation.   How would the repo know what the API pathing is?

Reply:
I don't think the repo should know what the API pathing is.  The name `NewCallbackUrl` to me suggested that this is the complete URL used for the callback, not that it is just a part of it and the documentation doesn't say either way as well.

Reply:
It may very well be the whole path... that's not been decided.  Once we decide what portion will be stored, I can update the godocs and add validation in the basic CallbackUrl type.